src/main.rs

```diff
@@ -119,140 +130,150 @@ fn process_buffer(buf: &[u8]) -> (HashMap<String, WeatherData>, u32) {
             }
 
             lines_count += 1;
-            station_name.clear();
+            station_name.fill(0);
             temperature = 0.0;
             negative_multiplier = 1;
             state = 0;
+            c = KEY_SIZE;
         }
-    }
-
-    return (station_temperatures, lines_count);
-}
 
-fn process_thread(
-    buf: &[u8],
-    extra_buffer_size: usize,
-    total_lines: Arc<Mutex<u32>>,
-) -> HashMap<String, WeatherData> {
-    let mut start_index = 0;
-    let mut end_index = buf.len() - extra_buffer_size;
-
-    for (index, &byte) in buf.iter().enumerate() {
-        if byte == b'\n' {
-            start_index = index + 1;
-            break;
+        if c + 1 < KEY_SIZE {
+            c += 1;
+        } else {
+            c = 0;
         }
-    }
-
-    for (index, &byte) in buf[buf.len() - extra_buffer_size..].iter().enumerate() {
-        if byte == b'\n' {
-            end_index = index + buf.len() - extra_buffer_size + 1;
-            break;
-        }
-    }
+    });
 
-    let (station_temperatures, lines_count) = process_buffer(&buf[start_index..end_index]);
-
-    let mut total_lines = total_lines.lock().unwrap();
-    *total_lines += lines_count;
-    drop(total_lines);
+    return (station_temperatures, lines_count);
+}
 
-    return station_temperatures;
+fn process_thread(buf: &[u8], extra_buffer_size: usize) -> (StationTemperatures, u32) {
+    let start_index = buf
+        .iter()
+        .position(|&b| b == b'\n')
+        .map(|i| i + 1)
+        .unwrap_or(0);
+
+    let buf_default_pos = buf.len() - extra_buffer_size;
+    let end_index = buf[buf_default_pos..]
+        .iter()
+        .position(|&b| b == b'\n')
+        .map(|i| i + buf_default_pos + 1)
+        .unwrap_or(buf_default_pos);
+
+    process_buffer(&buf[start_index..end_index])
 }
 
+const TOTAL_LINES: usize = 1_000_000_000;
+const AVG_ROW_SIZE: usize = 14;
+const THREAD_COUNT: usize = 250;
+const BUFFER_SIZE: usize = 2_000_000;
+const STAGE_COUNT: usize = (TOTAL_LINES * AVG_ROW_SIZE).div_ceil(THREAD_COUNT * BUFFER_SIZE);
+const SINGLE_ROW_SIZE: usize = 64;
```

Comment:
Great! I was thinking about how I could make this `STAGE_COUNT` variable dynamic.